### PR TITLE
Storing modelparts

### PR DESCRIPTION
**Pull request by tomkocmi**
When storing global_variables, also optimizer variables are stored and then when I want to apply modelpart initialization for only one of the parts and the rest of the model was changes, it breaks.

I have tried to find where the saving is happening but I didn't find any place in source codes.

### CODE REVIEW

**Comment by tomkocmi** [comment]
pylint opravim az spolu s pripadnymi reviews komentarema

File: neuralmonkey/model/model_part.py [hunk 1]
```diff
@@ -51,7 +51,7 @@ def feed_dict(self, dataset: Dataset, train: bool) -> FeedDict:
     def _init_saver(self) -> None:
         if not self._saver:
             parts_variables = tf.get_collection(
-                tf.GraphKeys.GLOBAL_VARIABLES, scope=self._variable_scope.name)
+                tf.GraphKeys.TRAINABLE_VARIABLES, scope=self._variable_scope.name)
```
Comment (jindrahelcl) [inline neuralmonkey/model/model_part.py]: opravdu? co když v tom scopu bude netrénovatelná variable, kterou si ta model part bude nastavovat?

File: neuralmonkey/learning_utils.py [hunk 2]
```diff
@@ -209,6 +209,13 @@ def training_loop(tf_manager: TensorFlowManager,
                                         tf_manager.best_score_batch),
                                 color='blue')
 
+                            # store also graph parts
+                            all_coders = set.union(
+                                *[rnr.all_coders for rnr in runners])
```
Comment (jindrahelcl) [inline neuralmonkey/learning_utils.py]: proč je tu ta transpozice?

File: neuralmonkey/learning_utils.py [hunk 3]
```diff
@@ -209,6 +209,13 @@ def training_loop(tf_manager: TensorFlowManager,
...
+                            for coder in all_coders:
+                                for session in tf_manager.sessions:
+                                    coder.save(session)
```
Comment (jindrahelcl) [inline neuralmonkey/learning_utils.py]: nejsem si jistej, jestli to chcem ukládat furt, @jlibovicky k tomu možná něco řikal a já to zapomněl. 

Navíc, jak je to tady teď, tak se to bude ukládat fakt furt, ne jen když je nejlepší model.

File: neuralmonkey/model/model_part.py (hunk 1 above)
Comment (jlibovicky) [inline neuralmonkey/model/model_part.py]: Přesně tak, pokud by tam byla schovaná batch nebo layer normalization, tak by se ty jejich počítané normalizační faktory neuložily a takovýhle bug bychom pak v životě nenašli.

File: neuralmonkey/learning_utils.py (hunk 3 above)
Comment (jlibovicky) [inline neuralmonkey/learning_utils.py]: Vyjadřoval jsem obavu, že je bude dost pomalé.

File: neuralmonkey/learning_utils.py (hunk 2 above)
Comment (jlibovicky) [inline neuralmonkey/learning_utils.py]: To jenom rozebere ten list na `*args` toho unionu. Nevím ale, jesti se v takovémhle případě nepíše ten generátor do obyčejných závorek.

Comment (jlibovicky) [inline neuralmonkey/learning_utils.py]: Navíc to sesbírání všech model parts se nemusí dít na tomhle místě, ale klidně úplně na začátku.

Další věc je, že bych nesbíral ty model parts jenom z runnerů, ale i z traineru (může být kus sítě, co se jenom trénuje, ale nikdy nepouští, proč ne), popřípadě jenom z traineru. Co se netrénuje, stejně není zajímavé ukládat.

File: neuralmonkey/model/model_part.py (hunk 1 above)
Comment (tomkocmi) [inline neuralmonkey/model/model_part.py]: kazdopadne GLOBAL_VARIABLES je take nepouzitelne protoze to uklada zase az moc a nejde to nacist samostatne bez napriklad optimizeru, coz predpokladam nechceme. Takze co s tim? ja se nevyznam ve vsech tech promenych, jen jsem testoval tuhle a ta ulozila vse pro danou konvolucni sit.

File: neuralmonkey/learning_utils.py (hunk 3 above)
Comment (tomkocmi) [inline neuralmonkey/learning_utils.py]: ted se to uklada vzdy, to jsem prehlednul. Kazdopadne nevim jake modely trenujete vy, ale myslim, ze kdyz se jednou za pul hodiny neco ulozi tak to nebude tak pomale, nebo to je fakt az tak pomale (predpokladal jsem tak do sekundy max nekolik sekund)

File: neuralmonkey/learning_utils.py (hunk 2 above)
Comment (tomkocmi) [inline neuralmonkey/learning_utils.py]: souhlasim, ze se musi ulozit vsechny model_parts, ale nevim jak to napsat elegantne nez tohle cele pak udelat jeste jednou pro trenery a propojit to. Navic to mozna neresi rekurzivni podobu (input>encoder>encoder>decoder>decoder>output) nekterych model_parts.

File: neuralmonkey/learning_utils.py (hunk 3 above)
Comment (jlibovicky) [inline neuralmonkey/learning_utils.py]: Je to několik málo sekund, ale to v celkové době trénování ztratí.

File: neuralmonkey/learning_utils.py (hunk 2 above)
Comment (jlibovicky) [inline neuralmonkey/learning_utils.py]: Runnery a trainery si sbírají všechny model parts rekurzivně. Potřebují to k tomu, aby mohli TF manageru říct, kam chtějí feedovat.

Sbíral bych spíš jenom z traineru. Smysl ukládat má jenom to, co se trénuje. Jinak, pokud tam mají být i ty části z runnerů, tak bych se s tím nepáral a do té list comprehension bych napsal `runners + [trainer]`.

File: neuralmonkey/model/model_part.py (hunk 1 above)
Comment (jlibovicky) [inline neuralmonkey/model/model_part.py]: Mně to přijde, že to má být `MODEL_VARIABLES`. Vizte https://www.tensorflow.org/api_docs/python/tf/GraphKeys

**Comment by tomkocmi** [comment]
Otestoval jsem MODEL_VARIABLES a funguji tak jak se ocekava. Je tu jedna vec, ze se to nyni uklada jen kdyz je lepsi nebo stejne dobry vysledek (aby to neukladalo na stejne dobre vysledky, tak bych musel asi rozvrtat tf_manager)

**Comment by jlibovicky** [comment]
To je tak málo pravděpodobný, že by bylo identický skóre, že bych to neřešil.

**Comment by jlibovicky** [comment]
Na řádku 205 dej zatím `# type: ignore`. (viz #415)

**Comment by jindrahelcl** [comment]
Zapomněl jsi to před mergem rebasnout
